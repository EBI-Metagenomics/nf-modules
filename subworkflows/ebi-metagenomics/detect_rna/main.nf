// General subwf to detect RNA by using provided RFAM models. It uses cmscan or cmsearch.
// Output: deoverlapped table and chosen fasta file with RNA sequences.

// Use cmscan mode if input fasta file is small and models file is quite big (usecase: mags-catalogues-pipeline)
// Important note: .cm file should be cmpress-ed before execution
// Use cmsearch mode if input fasta is massive and models file contains chosen set of models (usecase: ASA)


include { INFERNAL_CMSEARCH           } from '../../../modules/ebi-metagenomics/infernal/cmsearch/main'
include { INFERNAL_CMSCAN             } from '../../../modules/ebi-metagenomics/infernal/cmscan/main'
include { CONVERTCMSCANTOCMSEARCH     } from '../../../modules/ebi-metagenomics/convertcmscantocmsearch/main'
include { CMSEARCHTBLOUTDEOVERLAP     } from '../../../modules/ebi-metagenomics/cmsearchtbloutdeoverlap/main'
include { EASEL_ESLSFETCH             } from '../../../modules/ebi-metagenomics/easel/eslsfetch/main'


workflow DETECT_RNA {

    take:
    ch_fasta     // channel: [ val(meta), [ fasta ] ]
    rfam         // folder: rfam for cmsearch/cmscan
    claninfo     // file: claninfo for cmsearchtbloutdeoverlap
    mode         // cmsearch/cmscan

    main:

    ch_versions = Channel.empty()
    cmsearch_ch = Channel.empty()

    if ( mode == 'cmsearch' ) {
        INFERNAL_CMSEARCH(
            ch_fasta,
            rfam
        )
        ch_versions = ch_versions.mix(INFERNAL_CMSEARCH.out.versions.first())
        cmsearch_ch = INFERNAL_CMSEARCH.out.cmsearch_tbl
    }
    else if (mode == 'cmscan') {
       INFERNAL_CMSCAN(
            ch_fasta,
            rfam
       )
       ch_versions = ch_versions.mix(INFERNAL_CMSCAN.out.versions.first())

       CONVERTCMSCANTOCMSEARCH(INFERNAL_CMSCAN.out.cmscan_tbl)
       ch_versions = ch_versions.mix(CONVERTCMSCANTOCMSEARCH.out.versions.first())

       cmsearch_ch = CONVERTCMSCANTOCMSEARCH.out.cmsearch_tblout
    }

    CMSEARCHTBLOUTDEOVERLAP(
        cmsearch_ch,
        claninfo
    )
    ch_versions = ch_versions.mix(CMSEARCHTBLOUTDEOVERLAP.out.versions.first())

    ch_easel = ch_fasta
                .join(CMSEARCHTBLOUTDEOVERLAP.out.cmsearch_tblout_deoverlapped)
    EASEL_ESLSFETCH(
        ch_easel
    )
    ch_versions = ch_versions.mix(EASEL_ESLSFETCH.out.versions.first())

    emit:
    cmsearch_deoverlap_out = CMSEARCHTBLOUTDEOVERLAP.out.cmsearch_tblout_deoverlapped   // channel: [ val(meta), [ deoverlapped ] ]
    easel_out              = EASEL_ESLSFETCH.out.easel_coords                           // channel: [ val(meta), [ fasta ] ]
    versions               = ch_versions                                                // channel: [ versions.yml ]
}

