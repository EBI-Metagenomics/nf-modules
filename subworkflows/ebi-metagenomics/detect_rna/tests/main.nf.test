nextflow_workflow {

    name "Test Subworkflow DETECT_RNA"
    script "../main.nf"
    workflow "DETECT_RNA"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_ebimetagenomics"
    tag "subworkflows/detect_rna"
    tag "infernal"
    tag "infernal/cmsearch"
    tag "infernal/cmscan"
    tag "cmsearchtbloutdeoverlap"
    tag "convertcmscantocmsearch"
    tag "easel"
    tag "easel/eslsfetch"

    test("test_detect_rna cmsearch") {

        when {
            params {
                rfam = "${moduleDir}/tests/rfam"
                rfam_clan = "${moduleDir}/tests/ribo.claninfo"
            }
            workflow {
                """
                input[0] = Channel.of( [ [ id:'test', single_end:false ], // meta map
                    file('${moduleDir}/tests/test.fasta.gz', checkIfExists: true)
                ])
                input[1] = file(params.rfam)
                input[2] = file(params.rfam_clan)
                input[3] = "cmsearch"
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("test_detect_rna cmscan") {

        when {
            params {
                rfam = "${moduleDir}/tests/rfam"
                rfam_clan = "${moduleDir}/tests/ribo.claninfo"
            }
            workflow {
                """
                input[0] = Channel.of( [ [ id:'test', single_end:false ], // meta map
                    file('${moduleDir}/tests/test.fasta.gz', checkIfExists: true)
                ])
                input[1] = file(params.rfam)
                input[2] = file(params.rfam_clan)
                input[3] = "cmscan"
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
