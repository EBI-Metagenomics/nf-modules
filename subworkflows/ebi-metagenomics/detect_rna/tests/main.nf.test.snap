{
    "test_detect_rna cmscan": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tblout.deoverlapped:md5,eece0ccb016a5631dc33de1d0bb101d8"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_easel_coords.fasta:md5,83ce7b8c3957307d870ae865f3120bbf"
                    ]
                ],
                "2": [
                    "versions.yml:md5,1b7d07e444cf4802259240a7c038c1bc",
                    "versions.yml:md5,65638944226376bf03ace12b80a196a5",
                    "versions.yml:md5,a1ccd53172c56b58c83e08a977e0f3f6",
                    "versions.yml:md5,ee08705ca84ba0942119be25a75078c8"
                ],
                "cmsearch_deoverlap_out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tblout.deoverlapped:md5,eece0ccb016a5631dc33de1d0bb101d8"
                    ]
                ],
                "easel_out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_easel_coords.fasta:md5,83ce7b8c3957307d870ae865f3120bbf"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1b7d07e444cf4802259240a7c038c1bc",
                    "versions.yml:md5,65638944226376bf03ace12b80a196a5",
                    "versions.yml:md5,a1ccd53172c56b58c83e08a977e0f3f6",
                    "versions.yml:md5,ee08705ca84ba0942119be25a75078c8"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-21T18:30:52.849616"
    },
    "test_detect_rna cmsearch": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tblout.deoverlapped:md5,2ff422ba8581127bc95262d9751dbcba"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_easel_coords.fasta:md5,fb23f13526672a300845b767dbc5fd44"
                    ]
                ],
                "2": [
                    "versions.yml:md5,4bf517cd95bf099f8f0c13ef6735efcb",
                    "versions.yml:md5,65638944226376bf03ace12b80a196a5",
                    "versions.yml:md5,a1ccd53172c56b58c83e08a977e0f3f6"
                ],
                "cmsearch_deoverlap_out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tblout.deoverlapped:md5,2ff422ba8581127bc95262d9751dbcba"
                    ]
                ],
                "easel_out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_easel_coords.fasta:md5,fb23f13526672a300845b767dbc5fd44"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4bf517cd95bf099f8f0c13ef6735efcb",
                    "versions.yml:md5,65638944226376bf03ace12b80a196a5",
                    "versions.yml:md5,a1ccd53172c56b58c83e08a977e0f3f6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-21T18:19:23.851729"
    }
}